File: src/Core/Seguranca/Ataques/AtaqueBytesFixos.php
<?php

namespace Application\Core\Seguranca\Ataques;

use Application\Core\Seguranca\AlvoCriptografico;
use Application\Core\Seguranca\AtaqueInterface;
use Application\Core\Seguranca\ResultadoAtaque;

/**
 * As primeiras versões dessa cifra tinham um header fixo (43 bytes idênticos
 * sempre) ou um marcador constante (o '$' na posição 64). Esse ataque gera
 * vários tokens de textos diferentes e procura qualquer posição de byte que
 * NUNCA muda - isso é uma âncora que um atacante pode usar pra recuperar a
 * chave (como fizemos na primeira rodada dessa conversa).
 */
class AtaqueBytesFixos implements AtaqueInterface
{
    public function __construct(private int $amostras = 30)
    {
    }

    public function nome(): string
    {
        return 'Bytes fixos entre tokens';
    }

    public function executar(AlvoCriptografico $alvo): ResultadoAtaque
    {
        $tokens = [];
        for ($i = 0; $i < $this->amostras; $i++) {
            $texto = 'TEXTO_VARIADO_' . $i . str_repeat(chr(65 + ($i % 26)), $i % 10);
            $tokens[] = substr($alvo->encrypt($texto), strlen($alvo->prefixo()));
        }

        $decodificados = array_map(fn($t) => base64_decode(strtr($t, '-_', '+/') . str_repeat('=', (4 - strlen($t) % 4) % 4)), $tokens);
        $tamanhoMinimo = min(array_map('strlen', $decodificados));

        $posicoesFixas = [];
        for ($i = 0; $i < $tamanhoMinimo; $i++) {
            $valores = array_map(fn($d) => $d[$i], $decodificados);
            if (count(array_unique($valores)) === 1) {
                $posicoesFixas[] = $i;
            }
        }

        if (!empty($posicoesFixas)) {
            return new ResultadoAtaque(
                $this->nome(),
                vulneravel: true,
                severidade: 'alta',
                detalhes: count($posicoesFixas) . " posição(ões) de byte fixas em {$this->amostras} tokens: " . implode(', ', array_slice($posicoesFixas, 0, 10)),
                dados: ['posicoes' => $posicoesFixas],
            );
        }

        return new ResultadoAtaque(
            $this->nome(),
            vulneravel: false,
            severidade: 'info',
            detalhes: "0 de $tamanhoMinimo posições fixas em {$this->amostras} tokens",
        );
    }
}

File: packages/java/src/main/java/fbc/Seguranca/Ataques/AtaqueBytesFixos.java
package fbc.Seguranca.Ataques;

import fbc.Seguranca.AlvoCriptografico;
import fbc.Seguranca.AtaqueInterface;
import fbc.Seguranca.ResultadoAtaque;
import fbc.Seguranca.Severidade;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

/**
 * As primeiras versões dessa cifra tinham um header fixo ou um marcador
 * constante. Esse ataque gera vários tokens de textos diferentes e procura
 * qualquer posição de byte que NUNCA muda.
 */
public class AtaqueBytesFixos implements AtaqueInterface {
    private final int amostras;

    public AtaqueBytesFixos() {
        this(30);
    }

    public AtaqueBytesFixos(int amostras) {
        this.amostras = amostras;
    }

    @Override
    public String nome() {
        return "Bytes fixos entre tokens";
    }

    @Override
    public ResultadoAtaque executar(AlvoCriptografico alvo) {
        List<byte[]> decodificados = new ArrayList<>();
        for (int i = 0; i < this.amostras; i++) {
            String texto = "TEXTO_VARIADO_" + i
                    + String.valueOf((char) ('A' + (i % 26))).repeat(i % 10);
            String token = alvo.encrypt(texto);
            decodificados.add(alvo.base64urlDecode(token.substring(alvo.prefixo().length())));
        }

        int tamanhoMinimo = Integer.MAX_VALUE;
        for (byte[] d : decodificados) {
            tamanhoMinimo = Math.min(tamanhoMinimo, d.length);
        }

        List<Integer> posicoesFixas = new ArrayList<>();
        for (int i = 0; i < tamanhoMinimo; i++) {
            Set<Integer> valores = new HashSet<>();
            for (byte[] d : decodificados) {
                valores.add(d[i] & 0xFF);
            }
            if (valores.size() == 1) {
                posicoesFixas.add(i);
            }
        }

        if (!posicoesFixas.isEmpty()) {
            List<String> primeiras = new ArrayList<>();
            for (int i = 0; i < Math.min(10, posicoesFixas.size()); i++) {
                primeiras.add(String.valueOf(posicoesFixas.get(i)));
            }
            return new ResultadoAtaque(
                    nome(),
                    true,
                    Severidade.ALTA,
                    posicoesFixas.size() + " posição(ões) de byte fixas em " + this.amostras
                            + " tokens: " + String.join(", ", primeiras));
        }

        return new ResultadoAtaque(
                nome(),
                false,
                Severidade.INFO,
                "0 de " + tamanhoMinimo + " posições fixas em " + this.amostras + " tokens");
    }
}
